Add share attributes + prevent download permission

PVince81 opened the pull request:
Close https://github.com/nextcloud/server/issues/16413

- brings in share attributes column in oc_share
- store download permission in extended share attributes
- enforces download restriction

+ [x] backend
+ [x] frontend, add "hide download" permission
+ [ ] ~~extension points for secure view ?~~ not for now
+ [ ] BUG: share view-only file with link, open link then download => this should be blocked on backend level and in UI
+ [ ] BUG: share view-only folder with link, open link then download any entry => this should be blocked on backend level and in UI
+ [x] BUG: reshare with public link should automatically tick "hide download"
+ [x] BUG: reshare with public link should not allow to tick "hide download" 
+ [ ] BUG: viewer app still opens (optional)
+ [x] TEST: try increasing attribute permissions in reshare create
+ [x] TEST: try increasing attribute permissions in reshare update
+ [ ] TODO: make sure versions endpoint also doesn't allow download past versions

PVince81 committed:
Add share attributes + prevent download permission

Makes it possible to store download permission

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
regardless how many times I run `composer run cs:fix` locally it doesn't fix anything :rage:

PVince81 commented:
for testing, payload to set with curl on a share (here with id 2):
```
curl -u admin:admin 'https://vvortex.local/ocs/v2.php/apps/files_sharing/api/v1/shares/2' \
  -X 'PUT' \
  -H 'OCS-APIRequest: true' \
-H 'Accept: application/json, text/plain, */*' \
  -H 'Content-Type: application/json' \
  --data-raw '{"attributes":[{"scope": "permission", "key": "download", "enabled": "false"}]}'
```

PVince81 committed:
Fix share controller to accept share attributes alone

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Fix share attribute related tests + code style

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] consider renaming "enabled" to "value" as we likely want more than a boolean there for flexibility ? (also change the method interfaces then)

PVince81 commented:
sqlite only not reproducible locally :open_mouth: 
```
1) Test\Share20\DefaultShareProviderTest::testGetShareByIdUserShare
Invalid argument supplied for foreach()

/drone/src/lib/private/Share20/DefaultShareProvider.php:1563
/drone/src/lib/private/Share20/DefaultShareProvider.php:1082
/drone/src/lib/private/Share20/DefaultShareProvider.php:814
/drone/src/tests/lib/Share20/DefaultShareProviderTest.php:221
```

PVince81 committed:
Attempt to fix sqlite json issue

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
rebased for CI, hopefully everything green after that!

PVince81 commented:
- [ ] simplify permissions in DB ?
   - ex: `["permissions","download",false]` or even: `{'permissions': {'download': false}}`

PVince81 committed:
Don't reset share attributes when not specified

When not specified in the OCS call, don't reset the share attributes.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Add nc:share-attributes Webdav property

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Add share attrs + download permission support in frontend

Added download permission checkbox in frontend
Added share attributes parsing and setting in frontend.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've added a "Allow download" checkbox now in the frontend and also the logic to read and save it.
Additionally, the share attributes are now available in a "nc:share-attributes" DAV property for clients and is also used by the web UI.

More work needed to:
- [ ] make sure resharing cannot increase download perms
- [ ] make sure public link page doesn't show download action when resharing a received share that has no download perm

- [ ] see if possible to have JS tests for the new bits (at first look it seems we don't have any)

PVince81 commented on apps/files_sharing/lib/AppInfo/Application.php:
```diff
@@ -139,18 +143,76 @@ public function registerEventsScripts(IEventDispatcher $dispatcher, EventDispatc
+		$dispatcher->addListener(
+			'file.beforeGetDirect',
+			function (GenericEvent $event) use ($userSession, $rootFolder) {
+				$pathsToCheck[] = $event->getArgument('path');
```
move the square brackets to the right side

PVince81 commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -456,7 +462,8 @@ public function createShare(
 		string $note = '',
-		string $label = ''
+		string $label = '',
+		string $attributes = null
```
`?string`

PVince81 commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -1051,7 +1063,8 @@ public function updateShare(
 		string $label = null,
-		string $hideDownload = null
+		string $hideDownload = null,
+		string $attributes = null
```
`?string`

PVince81 commented on apps/files_sharing/lib/ViewOnly.php:
```diff
@@ -0,0 +1,116 @@
+	 * @param string[] $pathsToCheck
+	 * @return bool
+	 */
+	public function check($pathsToCheck) {
```
add strict type

PVince81 commented on apps/files_sharing/src/components/SharingEntry.vue:
```diff
@@ -390,6 +426,10 @@ export default {
+			if (this.share.hasDownloadPermission !== isDownloadChecked) {
+				this.share.hasDownloadPermission = isDownloadChecked
+				this.queueUpdate('attributes')
+			}
```
optional: check if we can also skip "permissions"

juliushaertl replied:
Would be good, seems with the current approach there are always two PUT requests. queueUpdate should allow to pass multiple properties to update: https://github.com/nextcloud/server/blob/e1a36bedb40f4ef4dfecc62cb7fac30256bba26f/apps/files_sharing/src/mixins/SharesMixin.js#L220

PVince81 commented on lib/private/legacy/OC_Files.php:
```diff
@@ -167,6 +169,14 @@ public static function get($dir, $files, $params = null) {
 			}
 
+			//Dispatch an event to see if any apps have problem with download
+			$event = new GenericEvent(null, ['dir' => $dir, 'files' => $files, 'run' => true]);
```
use custom class

PVince81 commented on lib/public/Share/IAttributes.php:
```diff
@@ -0,0 +1,68 @@
+namespace OCP\Share;
+
+/**
+ * Interface IAttributes
```
add message for "use only"

PVince81 commented on lib/public/Share/IShare.php:
```diff
@@ -314,6 +314,31 @@ public function setPermissions($permissions);
+	 * @since 25.0.0
+	 * @return IAttributes
+	 */
+	public function newAttributes();
```
add return type

PVince81 commented on lib/public/Share/IShare.php:
```diff
@@ -314,6 +314,31 @@ public function setPermissions($permissions);
+	 * @since 25.0.0
+	 * @return IShare The modified object
+	 */
+	public function setAttributes(?IAttributes $attributes);
```
add "use only" on the interface

juliushaertl commented on apps/dav/lib/DAV/ViewOnlyPlugin.php:
```diff
@@ -0,0 +1,114 @@
+	/**
+	 * @param ILogger $logger
+	 */
+	public function __construct(ILogger $logger) {
```
Switch to LoggerInterface?

juliushaertl commented on lib/private/legacy/OC_Files.php:
```diff
@@ -322,6 +338,19 @@ private static function getSingleFile($view, $dir, $name, $params) {
 			$rangeArray = self::parseHttpRangeHeader(substr($params['range'], 6), $fileSize);
 		}
 
+		$dispatcher = \OC::$server->query(IEventDispatcher::class);
+		$event = new GenericEvent(null, ['path' => $filename]);
+		$dispatcher->dispatch('file.beforeGetDirect', $event);
```
Could be a separate event class. There is already https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/apps/dav/lib/Events/BeforeFileDirectDownloadedEvent.php#L34 but that covers a different endpoint

juliushaertl commented:
Just had a first look at the code so far, not tested yet

- [ ] Probably needs some extra handling for the direct download endpoint that mobile clients may use (ref https://github.com/nextcloud/server/pull/9178)

PVince81 commented:
@juliushaertl any comments on the data structure for the permissions ? I have mixed feelings there as it could probably be simplified, at least outside the database.

Is this something that apps like richdocuments could easily reuse ?

PVince81 commented:
I've sent a separate PR just for the column migration so that we can have it in earlier: https://github.com/nextcloud/server/pull/32746

juliushaertl commented on apps/files_sharing/src/models/Share.js:
```diff
@@ -43,6 +43,14 @@ export default class Share {
+			} catch (e) {
+				console.warn('Could not parse share attributes returned by server: "' + ocsData.attributes + '"')
+			}
+		}
```
As the later parts assume that attributes is an array:

```suggestion
		} else {
			ocsData.attributes = []
		}
```

This makes sure that changing the attributes works properly in the frontend if none are present yet

juliusknorr committed:
Fix adding to empty attributes and duplicate request

Signed-off-by: Julius Härtl <jus@bitgrid.net>

PVince81 committed:
Fix view-only code after code review comments

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
- [ ] BUG: share view-only file with link, open link then download => this should be blocked
- [ ] BUG: reshare with public link should automatically tick "hide download"
- [ ] BUG: viewer app still opens

PVince81 commented:
- [ ] TODO: make sure versions endpoint also doesn't allow download past versions

PVince81 committed:
Block download when needed on direct download endpoint

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've fixed the DirectDownloadController to also fire the event, now it will return a 403 if a download is denied.

PVince81 committed:
Inherit hide download from share attributes

When resharing by link, if the download permission was removed through
share attributes, convert it to the hide download flag.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I managed to map the download permission to "hide_download" column at creation and update time

the UI still needs to be adjusted to prevent changing that permission

CarlSchwan commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -542,6 +550,14 @@ public function createShare(
 				throw new OCSNotFoundException($this->l->t('Public link sharing is disabled by the administrator'));
 			}
 
+			if ($share->getNode()->getStorage()->instanceOfStorage(SharedStorage::class)) {
```
Shouldn't this apply to all the share types? if one user share me a file, I shouldn't be able to share it with another user with more permissions

PVince81 replied:
yes. IIRC this might already be enforced on a different level

CarlSchwan committed:
Enforcing permission during resharing

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Multiple fixes

- Fix tests
- Use non deprecated event stuff
- Add a bit of type hinting to the new stuff
- More safe handling of instanceOfStorage (share might not be the first
  wrapper)
- Fix resharing

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

PVince81 commented:
:+1: code changes look fine

since I'm the PR submitter I can't approve it

@juliushaertl can you approve also ?

PVince81 commented:
@CarlSchwan please squash. and while you're at it, one of the new event files has warnings on the header so probably you can fix them in that go

PVince81 commented:
some more failures: https://drone.nextcloud.com/nextcloud/server/21540/9/4
```
1) OCA\DAV\Tests\Unit\DAV\Controller\DirectControllerTest::testGetUrlNonExistingFileId
Error: Object of class Mock_IURLGenerator_d72bc28a could not be converted to string

/drone/src/apps/dav/tests/unit/Controller/DirectControllerTest.php:87
```

reproducible locally

PVince81 commented:
please merge once CI passes!

CarlSchwan reviewed (approved):
ci issue unrelated

PVince81 commented:
/backport to stable24

elhananjair commented:
Hello, @PVince81 
I was waiting for this commit for so long, I tested it recently, and didn't work well with PDF files 😢. It works fine with other files I tested such as .docx .md. 

![Screenshot from 2022-08-17 20-23-40](https://user-images.githubusercontent.com/69236638/185211011-0ab08607-f85e-47e2-b8ae-3d603a90b9cd.png)
![Screenshot from 2022-08-17 20-50-31](https://user-images.githubusercontent.com/69236638/185211021-b66d7d92-d5ba-49e9-8d15-7ff1c3bce65b.png)

PVince81 commented:
@elhananjair yes, the different apps also need to be adjusted to properly recognize the permission and act accordingly.
for apps that don't support view only, they should simply not view the document at all

juliushaertl commented:
I filed https://github.com/nextcloud/files_pdfviewer/issues/649 about this

iMoshi commented:
So, I've finally tried the feature, and unchecked "Allow download" when sharing a folder with an internal user, and the user would get "Access forbidden" with basically any document including .pdf.

Is there something I'm missing inside the Administration Settings?

Also, can we have the option to "Allow download" when sharing a single file with an internal user?
